impulse/lfsr.nim

```diff
@@ -0,0 +1,233 @@
+  ##         in descending order. Exponent 0 can be omitted, since it is always
+  ##         implicitly added. For example, to use the `x^5 + x^3 + 1`
+  ##         polynomial you can set taps to `[5, 3]` or to `[5, 3, 0]` (or to
+  ##         `[5, 3].toTensor` or `[5, 3, 0].toTensor`).
+  ## - state: Initial state of the LFSR as an enum value (`single_true` or
+  ##          `all_true`). Defaults to `single_true`.
+  ## - verbose: Enable it to print additional logs
+  ## - counter_starts_at_zero: Start the count from 0 or 1 (defaults to `true`)
+  ##
+  ## Return:
+  ## - Ready to use LFSR object
+  when typeof(taps) is not Tensor:
+    let taps = taps.toTensor
+  let init_state = if state == all_true:
+    arraymancer.ones[bool](taps.max())
+  else:
+    zeros[bool](taps.max() - 1).append(true)
+  initLFSR(taps, conf = conf, state = init_state,
+    verbose = verbose, counter_starts_at_zero = counter_starts_at_zero)
+
+proc reset*(self: var LFSR) =
+  ## Reset the LFSR `state` and `count` to their initial values
+  self.state = self.init_state
+  self.count = 0
+
+proc next*(self: var LFSR, verbose = false,
+    store_sequence: static bool = false) : bool =
+  ## Run one cycle on LFSR with given feedback polynomial and
+  ## update the count, state, feedback bit, output bit and sequence
+  ##
+  ## Inputs:
+  ## - Preconfigured LFSR object
+  ## - verbose: Print additional logs even if LSRF.verbose is disabled
+  ## - store_sequence: static bool that enables saving the generated sequence
+  ##
+  ## Return:
+  ## - bool output bit
+  if self.verbose or verbose:
+    echo "State: ", self.state
+
+  if self.counter_starts_at_zero:
+    result = self.state[self.seq_bit_index]
+    when store_sequence:
+      self.sequence = self.sequence.append(result)
+
+  if self.conf == fibonacci:
+    var b = self.state[self.taps[0] - 1] xor self.state[self.taps[1] - 1]
+    if self.taps.size > 2:
+      for coeff in self.taps[2.._]:
+        b = self.state[coeff - 1] xor b
+
+    self.state = self.state.roll(1)
+    self.feedbackbit = b
+    self.state[0] = self.feedbackbit
+  else: # galois
+    self.feedbackbit = self.state[0]
+    self.state = self.state.roll(-1)
+    for k in self.taps[1.._]:
+      self.state[k-1] = self.state[k-1] xor self.feedbackbit
+
+  if not self.counter_starts_at_zero:
+    result = self.state[self.seq_bit_index]
+    when store_sequence:
+      self.sequence = self.sequence.append(result)
+
+  self.count += 1
+  self.outbit = result
+
+iterator generator*(lfsr: var LFSR,n: int,
+    store_sequence: static bool = false): bool =
+  ## Generator that will generate a random sequence of length `n`
+  ##
+  ## Inputs:
+  ## - Preconfigured LFSR object
+  ## - n: Number of random values to generate
+  ## - store_sequence: static bool that enables saving the generated sequence
+  ##
+  ## Return:
+  ## - Generated boolean values
+  yield lfsr.next(store_sequence = store_sequence)
```

Comment:
I don't quite understand why this is an `iterator`. The iterator only returns a single element. So it might as well just be a procedure? Calling it in a for loop will be redundant, no?

And it doesn't loop over `n` either so that `next` is called `n` times. Oversight or intended?

Reply:
I renamed the iterator to `items`, fixed the problem you explained and added a couple of new arguments to make it more flexible. Those new options might be overkill though :-). If you feel they are too much please let me know and I'll remove them.

Reply:
Looks good to me!